astropy/convolution/convolve.py

```diff
@@ -137,66 +137,83 @@ def convolve(array, kernel, boundary='fill', fill_value=0.,
     # but, if we just so happen to be lucky enough to have the input array
     # have exactly the desired type, we just alias to array_internal
 
-    # Check if kernel is kernel instance
-    if isinstance(kernel, Kernel):
-        # Check if array is also kernel instance, if so convolve and
-        # return new kernel instance
-        if isinstance(array, Kernel):
-            if isinstance(array, Kernel1D) and isinstance(kernel, Kernel1D):
-                new_array = convolve1d_boundary_fill(array.array, kernel.array,
-                                                     0, True)
-                new_kernel = Kernel1D(array=new_array)
-            elif isinstance(array, Kernel2D) and isinstance(kernel, Kernel2D):
-                new_array = convolve2d_boundary_fill(array.array, kernel.array,
-                                                     0, True)
-                new_kernel = Kernel2D(array=new_array)
-            else:
-                raise Exception("Can't convolve 1D and 2D kernel.")
-            new_kernel._separable = kernel._separable and array._separable
-            new_kernel._is_bool = False
-            return new_kernel
-        kernel = kernel.array
+    # If array and kernel are both Kernal instances, convolve here and return.
+    if isinstance(kernel, Kernel) and isinstance(array, Kernel):
+        if isinstance(array, Kernel1D) and isinstance(kernel, Kernel1D):
+            new_array = convolve1d_boundary_fill(array.array, kernel.array,
+                                                 0, True)
+            new_kernel = Kernel1D(array=new_array)
+        elif isinstance(array, Kernel2D) and isinstance(kernel, Kernel2D):
+            new_array = convolve2d_boundary_fill(array.array, kernel.array,
+                                                 0, True)
+            new_kernel = Kernel2D(array=new_array)
+        else:
+            raise Exception("Can't convolve 1D and 2D kernel.")
+        new_kernel._separable = kernel._separable and array._separable
+        new_kernel._is_bool = False
+        return new_kernel
 
-    # Check that the arguments are lists or Numpy arrays
+    # Copy or alias array to array_internal
     try:
-        # Note this won't copy if it doesn't have to -- which is okay
-        # because none of what follows modifies array_internal.
-        array_internal = np.asanyarray(array, dtype=float)
+        # Anything that's masked must be turned into NaNs for the interpolation.
+        # This requires copying. A copy is also needed for nan_treatment == 'fill'
+        # A copy prevents possible function side-effects of the input array.
+        if nan_treatment == 'fill' or np.ma.is_masked(array) or mask is not None:
+            if np.ma.is_masked(array):
+                # np.ma.maskedarray.filled() returns a copy.
+                array_internal = array_internal.filled(np.nan)
```

Comment:
I'm afraid so. It should be:
```python
array_internal = array.filled(np.nan)
```

I think. I'm just checking to see how such a school-boy was made in case there's something else at play here. Need to obviously add/change a test. They shouldn't have passed.

Reply:
yes, exactly - no test = master isn't technically broken, we just introduced a bug.... hmm... there's probably a name for that

Reply:
The test for masked input arrays is missing.

Reply:
Doh, even with the forward ref removed I broke it for masked ``int`` arrays. ``array.filled(np.nan)`` fails as it can't convert the ``NaN`` (a ``float``) to an ``int``.